### PR TITLE
[benchmark] Add benchmark for UnsafeMutableRawBufferPointer firstIndex(of:) and lastIndex(of:)

### LINKED ISSUE #63200: Implement `_customIndexOfEquatableElement` for `UnsafeRawBufferPointer`
`UnsafeRawBufferPointer` and `UnsafeMutableRawBufferPointer` perform element-by-element scans in `firstIndex(of: Element)` and `lastIndex(of: Element)`. It appears that these types should be amenable to some improvements, by using vectorized loops.

**Motivation**
`UnsafeRawBufferPointer.firstIndex(of:)` might be faster and/or more efficient with a specialized implementation.

**Solution**
Benchmarks should be created, and then implementations of the `Collection` customization points `_customIndexOfEquatableElement()` and `_customLastIndexOfEquatableElement()` should be investigated.

**Alternatives considered**
If no substantial improvements are found, then drop it. That fact should be documented so that someone else doesn't repeat this work in the future.

### PR DESCRIPTION
**Pull request by valeriyvan**

### CODE REVIEW

**Comment by glessard** [comment]
@swift-ci please benchmark

**Comment by glessard** [comment]
Benchmark looks reasonable. Do we really need to run it with 4 different orders of magnitude?

**Comment by glessard** [comment]
Benchmarking issue:
✅  | Benchmark Check Report
---|---
⚠️⏱ | `RawBuffer.1000.findFirst` execution took 3 μs. <br><sub> Increase the workload of `RawBuffer.1000.findFirst` to be more than 20 μs.

Removing the versions that take fewer than 10k elements should fix this.

**Review by glessard — CHANGES_REQUESTED** [review]
The benchmark harness needs each iteration to run for longer than some minimum amount of time. It appears that only the largest of these benchmark invocations meets the bar.

**Comment by valeriyvan** [comment]
> The benchmark harness needs each iteration to run for longer than some minimum amount of time. It appears that only the largest of these benchmark invocations meets the bar.

Search in 100K byte buffer looks to happen quite rare. Contrary, search in buffer of few hundred bytes (or shorter) looks more realistic. E.g. search for nil terminator in string buffers. Are we interested to keep an eye on how future code change might slow down search in short buffer? If yes, then we are interested to keep benchmarks of search in short buffers. Should we just increase multiplicator from 10 to 100 or 1000 and keep all benchmarks?

**Comment by glessard** [comment]
If you can increase the multiplier, then that would take care of the sanitizer's complaint. I'm not sure we need all those lengths, though. It matters to have a short length (less than 1 stride), a short length that vectorizes (more than 1 stride and less than 2 strides), and a large length (many strides.) Probably 7, 15 and 1000 are fine. We don't want to have an unnecessarily large number of benchmarks because of the load on the resources.

**Comment by glessard** [comment]
Created new issue to describe this work: https://github.com/apple/swift/issues/63200

**Comment by valeriyvan** [comment]
I don't know how review request from @lorentey got removed. It was unintentional. Sorry.

**Comment by glessard** [comment]
@swift-ci please benchmark

**Comment by glessard** [comment]
Looks like we can make the multipliers much larger:

| ✅    | Benchmark Check Report                                       |
| ---- | ------------------------------------------------------------ |
| ⚠️⏱   | `RawBuffer.7.findLast` execution took 3 μs. <br><sub> Increase the workload of `RawBuffer.7.findLast` to be more than 20 μs. <br><sub> unable to compute memory footprint of `RawBuffer.7.findLast` |
| ⚠️⏱   | `RawBuffer.7.findFirst` execution took 4 μs. <br><sub> Increase the workload of `RawBuffer.7.findFirst` to be more than 20 μs. <br><sub> unable to compute memory footprint of `RawBuffer.7.findFirst` <br><sub> unable to compute memory footprint of `RawBuffer.1000.findFirst` <br><sub> unable to compute memory footprint of `RawBuffer.1000.findLast` |
| ⚠️⏱   | `RawBuffer.15.findLast` execution took 5 μs. <br><sub> Increase the workload of `RawBuffer.15.findLast` to be more than 20 μs. <br><sub> unable to compute memory footprint of `RawBuffer.15.findLast` |
| ⚠️⏱   | `RawBuffer.15.findFirst` execution took 6 μs. <br><sub> Increase the workload of `RawBuffer.15.findFirst` to be more than 20 μs. <br><sub> unable to compute memory footprint of `RawBuffer.15.findFirst` |

**Comment by valeriyvan** [comment]
> Since we expect to speed up the search a lot, we can probably shoot for the millisecond range, so that the benchmark still executes for long enough after the vectorized implementations are in place.

Increased multiplier to 1000_000

**Comment by glessard** [comment]
@swift-ci please benchmark

**Comment by glessard** [comment]
@swift-ci please benchmark

File: benchmark/single-source/BufferFind.swift [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+func bufferSetup(size: Int, alignment: Int) {
+  buffer = UnsafeMutableRawBufferPointer.allocate(byteCount: size, alignment: alignment)
+  buffer.initializeMemory(as: UInt8.self, repeating: UInt8.min)
+  buffer[size / 2] = UInt8.max
+}
+
+@inline(never)
+public func run_BufferFindFirst(n: Int) {
+  var offset = 0
+  for _ in 0 ..< n * 1000_000 {
```
Comment (glessard) [inline benchmark/single-source/BufferFind.swift:91]: Let's make this multiplier 10_000

File: benchmark/single-source/BufferFind.swift [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      offset += index
+    }
+  }
+  blackHole(offset)
+}
+
+@inline(never)
+public func run_BufferFindLast(n: Int) {
+  var offset = 0
+  for _ in 0 ..< n * 1000_000 {
```
Comment (glessard) [inline benchmark/single-source/BufferFind.swift:102]: Let's make this multiplier 10_000

**Review by glessard — CHANGES_REQUESTED** [review]

File: benchmark/single-source/BufferFind.swift (hunk 2 above)
Comment (valeriyvan) [reply benchmark/single-source/BufferFind.swift:102]: done

File: benchmark/single-source/BufferFind.swift (hunk 1 above)
Comment (valeriyvan) [reply benchmark/single-source/BufferFind.swift]: done

Comment (glessard) [reply benchmark/single-source/BufferFind.swift]: Here was the output from the benchmark runner with the 1_000_000 multiplier:


| ✅    | Benchmark Check Report                                       |
| ---- | ------------------------------------------------------------ |
| ⚠️⏱   | `RawBuffer.7.findFirst` execution took at least 4335 μs. <br><sub> Decrease the workload of `RawBuffer.7.findFirst` by a factor of 8 (10), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `RawBuffer.7.findFirst` |
| ⛔️⏱   | `RawBuffer.1000.findLast` execution took at least 282237 μs. <br><sub> Decrease the workload of `RawBuffer.1000.findLast` by a factor of 512 (1000), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `RawBuffer.1000.findLast` |
| ⚠️⏱   | `RawBuffer.15.findLast` execution took at least 5665 μs. <br><sub> Decrease the workload of `RawBuffer.15.findLast` by a factor of 8 (10), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `RawBuffer.15.findLast` |
| ⚠️⏱   | `RawBuffer.7.findLast` execution took at least 3486 μs. <br><sub> Decrease the workload of `RawBuffer.7.findLast` by a factor of 4 (10), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `RawBuffer.7.findLast` |
| ⚠️⏱   | `RawBuffer.15.findFirst` execution took at least 7626 μs. <br><sub> Decrease the workload of `RawBuffer.15.findFirst` by a factor of 8 (10), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `RawBuffer.15.findFirst` |
| ⛔️⏱   | `RawBuffer.1000.findFirst` execution took at least 448424 μs. <br><sub> Decrease the workload of `RawBuffer.1000.findFirst` by a factor of 512 (1000), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `RawBuffer.1000.findFirst` |

**Review by glessard — APPROVED** [review]
Thanks for this benchmark! We can tune the remaining details once we have an implementation.

The benchmark checker still warns that the 1000-length iterations are slightly too long, but if the improvements are reasonable, then the result will be under the warning threshold.

**Comment by glessard** [comment]
@swift-ci please smoke test and merge

**Comment by glessard** [comment]
@swift-ci please smoke test macOS platform

**Comment by glessard** [comment]
@swift-ci please smoke test and merge

**Comment by glessard** [comment]
@swift-ci please smoke test and merge

**Comment by glessard** [comment]
@swift-ci please smoke test

**Pull request merged**